lib/vector-core/src/stream/batcher.rs

```diff
@@ -0,0 +1,540 @@
+where
+    I: ByteSizeOf,
+{
+    /// Create a new Batch instance
+    ///
+    /// Creates a new batch instance with specific element and allocation
+    /// limits. The element limit is a maximum cap on the number of `I`
+    /// instances. The allocation limit is a soft-max on the number of allocated
+    /// bytes stored in this batch, not taking into account overhead from this
+    /// structure itself. Caller is responsible for ensuring that `I` will fit
+    /// inside the allocation limit.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the allocation limit will not store at least
+    /// 1 instance of `I`, as measured by `mem::size_of`.
+    fn new(element_limit: usize, allocation_limit: usize) -> Self {
+        assert!(allocation_limit >= mem::size_of::<I>());
+        Self {
+            allocated_bytes: 0,
+            element_limit,
+            allocation_limit,
+            elements: Vec::with_capacity(element_limit),
+        }
+    }
+
+    /// Unconditionally insert an element into the batch
+    ///
+    /// This function is similar to `push` except that the caller does not need
+    /// to call `has_space` prior to calling this and it will never
+    /// panic. Intended to be used only when insertion must not fail.
+    fn with(mut self, value: I) -> Self {
+        self.allocated_bytes += value.size_of();
+        self.elements.push(value);
+        self
+    }
+
+    /// Decompose the batch
+    ///
+    /// Called by the user when they want to get at the internal store of
+    /// items. Returns a tuple, the first element being the allocated size of
+    /// stored items and the second the store of items.
+    fn destruct(self) -> (usize, Vec<I>) {
+        (self.allocated_bytes, self.elements)
+    }
+
+    /// Whether the batch has space for a new item
+    ///
+    /// This function returns true of there is space both in terms of item count
+    /// and byte count for the given item, false otherwise.
+    fn has_space(&self, value: &I) -> bool {
+        let too_many_elements = self.elements.len() + 1 > self.element_limit;
+        let too_many_bytes = self.allocated_bytes + value.size_of() > self.allocation_limit;
+        !(too_many_elements || too_many_bytes)
+    }
+
+    /// Push an element into the batch
+    ///
+    /// This function pushes an element into the batch. Callers must be sure to
+    /// call `has_space` prior to calling this function and receive a positive
+    /// result.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if there is not sufficient space in the batch
+    /// for a new element to be inserted.
+    fn push(&mut self, value: I) {
+        assert!(self.has_space(&value));
+        self.allocated_bytes += value.size_of();
+        self.elements.push(value);
+    }
+}
+
+#[pin_project]
+pub struct Batcher<St, Prt, T>
+where
+    Prt: Partitioner,
+{
+    /// The timer to maintain periodic flushes from this `Batcher`
+    timer: T,
```

Comment:
Should there not be a timer for each batch depending on when they were first created?

Reply:
Maybe? So the previous batching mechanism had the semantics that every partition batch would be flushed every N seconds with a clock that started on first insertion of the partition key. This mechanism proposes a method that flushes at least every N seconds with a global clock. The upside of this approach is there's less mechanism overall -- no heap to maintain -- with the downside that batches will sometimes flush that might otherwise have been able to accumulate for a spell. In terms of when batches would have otherwise flushed this method makes batches emit along a common time boundary, per-batch timers potentially spreads the flushes out even if there's data that could be otherwise processed. 

So far as I can tell the previous mechanism is not documented to have a firm deadline and is best effort. This preserves those guarantees with less overhead and a different distribution to downstream consumers.

Reply:
I agree having a timer per batch would lead to more complex code. I think it makes more sense for communication efficiency, but we'll see where this goes.